File: leetcode/238-product-of-array-except-self/Solution.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> output(n); 

        output[0] = 1;
        for(int i = 1; i < n; i++) {
            output[i] = output[i-1] * nums[i-1];
        }

        int post = nums[n-1];
        for(int i = n-2; -1 < i; i--) {
            output[i] *= post;
            post *= nums[i];
        }


        return output;
    }
};

File: leetcode/238-product-of-array-except-self/Solution.py
#
# @lc app=leetcode id=238 lang=python3
#
# [238] Product of Array Except Self
#
# @lc code=start
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        k = [0] * len(nums)
        p = 1
        nan = -1

        for idx, n in enumerate(nums):
            if n == 0:
                if nan == -1:
                    nan = idx
                else:
                    return [0] * len(nums)
            else:
                p *= n


        for idx, n in enumerate(nums):
            if nan == -1:
                k[idx] = int(p/n)
            else:
                k[idx] = 0 if idx != nan else p 
            

            
        return k
# @lc code=end
